sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala

```diff
@@ -373,284 +191,276 @@ case class SumDistinct(child: Expression)
-  extends AggregateFunction {
-  def this() = this(null, null, 0) // Required for serialization.
-
-  private val hyperLogLog = new HyperLogLog(relativeSD)
-
-  override def update(input: Row): Unit = {
-    val evaluatedExpr = expr.eval(input)
-    hyperLogLog.addAll(evaluatedExpr.asInstanceOf[HyperLogLog])
+  override def merge(value: Row, rowBuf: MutableRow): Unit = {
+    if (!value.isNullAt(aggr.ordinal)) {
+      arg.value = value(aggr.ordinal)
+      buffer.value = rowBuf(aggr.ordinal)
+      if (rowBuf.isNullAt(aggr.ordinal) || cmp.eval(null) == true) {
+        rowBuf.update(aggr.ordinal, arg.value)
+      }
+    }
   }
 
-  override def eval(input: Row): Any = hyperLogLog.cardinality()
+  override def terminate(row: Row): Any = aggr.eval(row)
 }
 
-case class SumFunction(expr: Expression, base: AggregateExpression) extends AggregateFunction {
-  def this() = this(null, null) // Required for serialization.
+case class CountFunction(aggr: BoundReference, base: Count)
+    extends AggregateFunction {
+  override def reset(buf: MutableRow): Unit = {
+    buf.update(aggr.ordinal, 0L)
+  }
 
-  private val calcType =
-    expr.dataType match {
-      case DecimalType.Fixed(_, _) =>
-        DecimalType.Unlimited
-      case _ =>
-        expr.dataType
+  override def iterate(argument: Any, buf: MutableRow): Unit = {
+    if (argument != null) {
+      if (buf.isNullAt(aggr.ordinal)) {
+        buf.setLong(aggr.ordinal, 1L)
+      } else {
+        buf.update(aggr.ordinal, buf.getLong(aggr.ordinal) + 1L)
+      }
     }
-
-  private val zero = Cast(Literal(0), calcType)
-
-  private val sum = MutableLiteral(null, calcType)
-
-  private val addFunction = Coalesce(Seq(Add(Coalesce(Seq(sum, zero)), Cast(expr, calcType)), sum))
-
-  override def update(input: Row): Unit = {
-    sum.update(addFunction, input)
   }
 
-  override def eval(input: Row): Any = {
-    expr.dataType match {
-      case DecimalType.Fixed(_, _) =>
-        Cast(sum, dataType).eval(null)
-      case _ => sum.eval(null)
+  override def merge(value: Row, rowBuf: MutableRow): Unit = {
+    if (value.isNullAt(aggr.ordinal)) {
+      // do nothing
+    } else if (rowBuf.isNullAt(aggr.ordinal)) {
+      rowBuf(aggr.ordinal) = value(aggr.ordinal)
+    } else {
+      rowBuf.update(aggr.ordinal, value.getLong(aggr.ordinal) + rowBuf.getLong(aggr.ordinal))
     }
   }
+
+  override def terminate(row: Row): Any = aggr.eval(row)
 }
 
-case class SumDistinctFunction(expr: Expression, base: AggregateExpression)
+case class CountDistinctFunction(aggr: BoundReference, base: CountDistinct)
   extends AggregateFunction {
-
-  def this() = this(null, null) // Required for serialization.
-
-  private val seen = new scala.collection.mutable.HashSet[Any]()
-
-  override def update(input: Row): Unit = {
-    val evaluatedExpr = expr.eval(input)
-    if (evaluatedExpr != null) {
-      seen += evaluatedExpr
+  override def reset(buf: MutableRow): Unit = {
+    buf.update(aggr.ordinal, 0L)
```

Comment:
Style nit:  `buf(agg.ordinal) = 0L`

We might consider adding `def update(ref: BoundReference, newValue: Any)` also as sugar.